posts/2023-04-13-rust-by-examples.adoc

```diff
@@ -0,0 +1,726 @@
+[source,rust]
+----
+pub struct MyStruct {...}
+pub enum MyEnum {...}
+pub fn my_func() {...}
+----
+
+=== Ownership
+
+Ownership is a set of rules that govern how a Rust program manages memory. If any of the rules is violated, the program won't compile:
+
+* Each value in Rust has an owner.
+* There can only be one owner at a time.
+* When the owner goes out of scope, the value will be dropped.
+
+This is not an issue for primitive types, because they are small, and they are copied as an argument or a return value. Passing on non-primitive types to a function will move their ownership to the function and this ownership is not returned. For example:
+[source,rust]
+----
+fn return_match(pattern: String, lines: Vec<&str>) -> Option<String> {
+    lines
+        .iter()
+        .find(|&line| line.contains(&pattern))
+        .map(|&line| line.to_string())  // map &str to a String instance
+}
+
+fn example_2() {
+    let lines = vec!["abcde", "defgh", "ghijk"];
+    let pattern = "gh".to_string();
+    let line = return_match(pattern, lines);
+    // at this point "lines" and "pattern" went out of scope,
+    // their ownership is transferred to the return_match() function
+}
+----
+
+If we want to use these parameters again, we could pass their copy as argument:
+[source,rust]
+----
+    let line = return_match(pattern.clone(), lines.clone());
+----
+
+However, cloning large values is expensive and clone() is not necessarily implemented for all custom types.
+The solution is to pass non-primitive types as references. The &x syntax lets us create a reference that refers to the value of x but does not own it. Because it does not own it, the value it points to will not be dropped when the reference stops being used. We call the action of creating a reference borrowing. As in real life, if a person owns something, you can borrow it from them. When you're done, you have to give it back. You don't own it.
+
+We could just return the found &str, and save the creation of the String, but the following code will fail to compile:
+[source,rust]
+----
+fn return_match_borrow(pattern: &str, lines: &Vec<&str>) -> Option<&str> {
+    lines
+        .iter()
+        .find(|&line| line.contains(&pattern))
+        .map(|&line| line)
+}
+// error: missing lifetime specifier
+// this function's return type contains a borrowed value, but the signature does not say whether it is borrowed from `pattern` or one of `lines`'s 2 lifetimes
+----
+
+We'll need to define the lifetime of the response and bind it to the lifetime of a function argument. Lifetimes are defined as labels in the format of ``'x``, where x identifies the lifetime:
+[source,rust]
+----
+fn return_match_borrow<'a>(pattern: &str, lines: &'a Vec<&str>) -> Option<&'a str> {
+....
+----
+
+We have a special case, our input are static values and we can also use the special lifetime label ``'static``:
+[source,rust]
```

Comment:
Maybe make it more clear that the static case is separate; Something along the lines of:

"In this example, we have a special case that allows us to handle the borrowing in a different way. Our inputs are static values ..."

Reply:
I explained the `static` case a bit better now